eq_translations/survey_schema.py

```diff
@@ -166,80 +219,113 @@ def get_catalog(self):
         total_translations = len(self.no_context_pointers + self.context_pointers)
 
         for pointer in self.no_context_pointers:
-            pointer_contents = resolve_pointer(self.schema, pointer)
-            if pointer_contents:
-                catalog.add(dumb_to_smart_quotes(pointer_contents))
+            self._add_message_to_catalog(pointer, catalog, with_context=False)
 
         for pointer in self.context_pointers:
-            pointer_contents = resolve_pointer(self.schema, pointer)
-            if pointer_contents:
-                parent_answer_id = self.get_parent_id(pointer)
-                question = self.get_parent_question(pointer)
-
-                message_context = "Answer for: {}".format(
-                    question["text"] if is_placeholder(question) else question
-                )
-
-                catalog.add(
-                    dumb_to_smart_quotes(pointer_contents),
-                    context=message_context,
-                    auto_comments=[f"answer-id: {parent_answer_id}"],
-                )
+            self._add_message_to_catalog(pointer, catalog, with_context=True)
 
         print(f"Total Messages: {total_translations}")
 
         return catalog
 
-    def translate(self, schema_translation):
+    def translate(self, schema_translation, language_code):
         """
-        Use the supplied schema translation object to translate pointers found
+        Use the supplied schema translation object and language code to translate pointers found
         within the survey
 
-        :param schema_translation:
+        :param schema_translation: The SchemaTranslation object
+        :param language_code: The target language code
         :return:
         """
         translated_schema = copy.deepcopy(self.schema)
         total_translations = len(self.no_context_pointers + self.context_pointers)
         missing_translations = 0
 
         for pointer in self.no_context_pointers:
-            pointer_contents = resolve_pointer(self.schema, pointer)
-            if pointer_contents:
-                translation = schema_translation.translate_message(pointer_contents)
-                if translation:
-                    translated_schema = set_pointer(
-                        translated_schema, pointer, translation
-                    )
-                else:
-                    missing_translations += 1
-                    print(
-                        "Missing translation at {}: '{}'".format(
-                            pointer, pointer_contents
-                        )
-                    )
+            missing_translations += self._handle_translation_for_pointer(
```

Comment:
I think this would be better as an immutable method which takes the entire list of pointers and returns a copy of the translated schema.

Reply:
👍